"""

Quast options:

--output-dir outdir
-R reference_genome
--genes genes.{gff,bed,txt,ncbi}
--operons operons.{gff,bed,txt,ncbi}
--min-contig lens (500)
--threads 
--labels comma,separated
-L labels from dirname
--bam filename


"""


rule quast:
    message:
        "QUAST qc'ing {wildcards.dir} assemblies"
    input:
        "{dir}/{:targets:}.contigs.fasta.gz"
    output:
        report="{dir}_qc/report.txt",
        dir="{dir}_qc/"
    conda:
        srcdir("quast.yml")
    threads:
        8
    shell:"""
    quast -o {output.dir} \
          -t {threads} \
          --meta \
          -L \
          {input}
    """


rule quast_report:
    message: "MultiQC aggregating QUAST reports"
    input:
        "{dir}_qc/report.txt"
    output:
        "reports/{dir}_qc.html"
    params:
        dir = "{dir}_qc"
    conda:
        srcdir("multiqc.yml")
    threads:
        1
    shell: """
    multiqc -d -m quast --no-data-dir {params.dir} -n {output}
    """
